utils/rate.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// EDIT: slight modification to allow setting rate limit on the fly
+// SCOPE: LeakyBucket
+package utils
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+type LeakyBucket struct {
+	//lint:ignore U1000 Padding is unused but it is crucial to maintain performance
+	// of this rate limiter in case of collocation with other frequently accessed memory.
+	prepadding [64]byte // cache line size = 64; created to avoid false sharing.
+	state      int64    // unix nanoseconds of the next permissions issue.
+	//lint:ignore U1000 like prepadding.
+	postpadding [56]byte // cache line size - state size = 64 - 8; created to avoid false sharing.
+
+	perRequest time.Duration
+	maxSlack   time.Duration
+	clock      Clock
+}
+
+func NewLeakyBucket(rateLimit int, slack time.Duration, clock Clock) *LeakyBucket {
+	var lb LeakyBucket
+	lb.SetRateLimit(rateLimit)
+	lb.maxSlack = slack * lb.perRequest
+	lb.clock = clock
+	atomic.StoreInt64(&lb.state, 0)
+	return &lb
+}
+
+func (lb *LeakyBucket) SetRateLimit(rateLimit int) {
+	lb.perRequest = time.Second / time.Duration(rateLimit)
```

Comment:
are `SetRateLimit` and `Take` meant to be externally synchronized? the go race detector will raise errors about this. we don't seem to have a test for this

Reply:
Yes, there were meant to be externally synchronized. But, rethinking about this, it would make more sense to handle the synchronization here, especially `Take` is already synchronized. I'll work on this.

Reply:
Spent too much time to make this work with atomic. At the end, the code became complicated since `state` has occupied 64 bits. It's possible to cramp the rate limit state to the 64 bits, but it has to be u64. Then, handling time without negative numbers became tricky... 

At the end, I reverted to the good ol' mutex. For now, we're not going to run into a performance issue since we only use this in QoS from a single goroutine. In the future, we might to revisit if there's a need for large amount of goroutines accessing this.

Reply:
@paulwe added a test